src/sentry/models/outbox.py

```diff
@@ -524,28 +521,45 @@ def save(self, **kwds: Any) -> None:  # type: ignore[override]
         metrics.incr("outbox.saved", 1, tags=tags)
         super().save(**kwds)
 
+    def lock_id(self, attrs: Iterable[str]) -> int:
+        # 64 bit integer that roughly encodes a unique, serializable lock identifier
+        return mmh3.hash64(".".join(str(getattr(self, attr)) for attr in attrs))[0]
+
     @contextlib.contextmanager
     def process_shard(
         self, latest_shard_row: OutboxBase | None
     ) -> Generator[OutboxBase | None, None, None]:
         flush_all: bool = not bool(latest_shard_row)
-        next_shard_row: OutboxBase | None
         using: str = db.router.db_for_write(type(self))
-        with transaction.atomic(using=using), django_test_transaction_water_mark(using=using):
-            try:
-                next_shard_row = (
-                    self.selected_messages_in_shard(latest_shard_row=latest_shard_row)
-                    .select_for_update(nowait=flush_all)
```

Comment:
Kind of.  To be honest, since all rows belong to the same shard here, any lock on any of these rows effectively locks the shard (because rows are only deleted in the same transaction as the locking). 

Now, none of that is necessary because the locking is explicit on the hash, and the deletion of rows need not be a transaction as the lock is known to be acquired at the time anyways.

In practice, the serialization here is strong and there won't be any case of multiple workers acting on the same rows.

Reply:
To be clear, "session based" does not mean it isn't shared across workers -- it only means that the lock is released automatically if the connection is dropped.  that's it.

Reply:
And also this is already tested in the `test_outbox` where I used a separate thread to force a separate connection and validate the serialization when both access the same shard.

Reply:
> To be clear, "session based" does not mean it isn't shared across workers -- it only means that the lock is released automatically if the connection is dropped. that's it.

:+1: for some reason I thought misinterpreted 'session based' with transaction scoped locks, but they are different lock scopes.